src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -40,131 +42,207 @@
+    @ForceInline
+    private static String readByte(AbstractMemorySegmentImpl segment, long offset, Charset charset) {
+        final int len = strlenByte(segment, offset, segment.byteSize());
+        final byte[] bytes = new byte[len];
+        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, len);
         return new String(bytes, charset);
     }
 
-    private static void writeByte(MemorySegment segment, long offset, Charset charset, String string) {
+    @ForceInline
+    private static void writeByte(AbstractMemorySegmentImpl segment, long offset, Charset charset, String string) {
         int bytes = copyBytes(string, segment, charset, offset);
         segment.set(JAVA_BYTE, offset + bytes, (byte)0);
     }
 
-    private static String readShort(MemorySegment segment, long offset, Charset charset) {
-        long len = chunkedStrlenShort(segment, offset);
-        byte[] bytes = new byte[(int)len];
-        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, (int)len);
+    @ForceInline
+    private static String readShort(AbstractMemorySegmentImpl segment, long offset, Charset charset) {
+        int len = strlenShort(segment, offset, segment.byteSize());
+        byte[] bytes = new byte[len];
+        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, len);
         return new String(bytes, charset);
     }
 
-    private static void writeShort(MemorySegment segment, long offset, Charset charset, String string) {
+    @ForceInline
+    private static void writeShort(AbstractMemorySegmentImpl segment, long offset, Charset charset, String string) {
         int bytes = copyBytes(string, segment, charset, offset);
         segment.set(JAVA_SHORT_UNALIGNED, offset + bytes, (short)0);
     }
 
-    private static String readInt(MemorySegment segment, long offset, Charset charset) {
-        long len = strlenInt(segment, offset);
-        byte[] bytes = new byte[(int)len];
-        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, (int)len);
+    @ForceInline
+    private static String readInt(AbstractMemorySegmentImpl segment, long offset, Charset charset) {
+        int len = strlenInt(segment, offset, segment.byteSize());
+        byte[] bytes = new byte[len];
+        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, len);
         return new String(bytes, charset);
     }
 
-    private static void writeInt(MemorySegment segment, long offset, Charset charset, String string) {
+    @ForceInline
+    private static void writeInt(AbstractMemorySegmentImpl segment, long offset, Charset charset, String string) {
         int bytes = copyBytes(string, segment, charset, offset);
         segment.set(JAVA_INT_UNALIGNED, offset + bytes, 0);
     }
 
     /**
-     * {@return the shortest distance beginning at the provided {@code start}
-     *  to the encountering of a zero byte in the provided {@code segment}}
+     * {@return the shortest distance beginning at the provided {@code fromOffset}
```

Comment:
Nit: instead of saying "return shortest distance" perhaps we should say "return the offset of the first zero byte" ?